src/lib/logical_query_plan/union_node.cpp

```diff
@@ -19,9 +19,12 @@ std::string UnionNode::description(const DescriptionMode mode) const {
 }
 
 std::vector<std::shared_ptr<AbstractExpression>> UnionNode::output_expressions() const {
-  Assert(expressions_equal(left_input()->output_expressions(), right_input()->output_expressions()),
-         "Input Expressions must match");
-  return left_input()->output_expressions();
+  const auto& left_expressions = left_input()->output_expressions();
+  if (set_operation_mode != SetOperationMode::Positions) {
+    Assert(expressions_equal(left_expressions, right_input()->output_expressions()),
+           "Input Expressions must match");
+  }
+  return left_expressions;
```

Comment:
Any idea on why we haven't had the relaxed case for Union Positions before?

Reply:
One could see this as an additional sanity check: Union Positions should have the same source and only different Predicates operating on them, so these expressions need to be the same. Doing more stuff is a violation of the concept of out Union Positions (that we might want to fail?). So that's why this is somehow the most arguable change. IMO there are three options:

1. keep this code as it was and always do a sanity check, 
2. keep this code as it is now and only do the sanity check if it is not a Union Positions (maybe even with `DebugAssert`).
3. keep the sanity checks in any case, but use `DebugAssert`.

Reply:
> Doing more stuff is a violation of the concept of our Union Positions

In that case I would would rule out option 2.

Reply:
Okay, what about a `DebugAssert(...)` for all Union Modes?

Reply:
Update here: Out of curiosity, I looked on what the `UnionPositions` operator does. It has a `DebugAssert` to check whether left and right input table have the same columns, so maybe a `DebugAssert` is also sufficient here (the check here is more costly, as the expressions need to be calculated, while the table column definitions are already initialized while executing the operator)

```c++
std::shared_ptr<const Table> UnionPositions::_prepare_operator() {
  DebugAssert(left_input_table()->column_definitions() == right_input_table()->column_definitions(),
              "Input tables don't have the same layout");
...
```

Reply:
I would also stick with the `Assert` here. That way, we fail as early as possible.

Reply:
I'll make a quick experiment on `Assert`/`DebugAssert` in `UnionNode::output_expressions()`. Regarding performance impact, keep in mind that the `Assert` causes fetching the expressions of both inputs, adding a call to the original subplan of (multiple) `UnionPosition` nodes for each `UnionNode`.

Reply:
I tested both versions on pella with TPC-H SF 0.01 and 10, TPC-DS SF 10, and JOB. There were some variations, but usually in queries with no `UnionNode`. I guess this is caused by the fact that for almost all plans containing union nodes, these nodes are pushed right after table scans that have the respecting stored table nodes as input. So leaving the `Assert` seems to be a good idea, but I would argue that switching the `DebugAssert` in the union operators to `Assert` should be done, too. Maybe we can add a comment in `UnionNode::output_expressions()` that explains this decision. There could be performance implications in the future if union nodes are placed elsewhere, e.g., if there are more queries that union multiple short-running select statements.

Such a comment could look like the following:
_Asserting matching table schemas leads to multiple fetches of a subplan's output expressions. Though this does not have performance implications now, they may arise in the future._

Reply:
Okay, let's do it like that. Maybe add to the comment that we changing the `Assert` to `DebugAssert` should be considered in the future.